Dockerfile

```
@@ -0,0 +1,19 @@
+FROM golang:1.19.3-alpine3.16
+
+ENV ROOT=/go/src/github.com/ha-zu/learn-clean-dev
+ENV MAIN=${ROOT}/src/cmd/cleanarchitecture
+ENV BUILD=${ROOT}/src/build
+WORKDIR ${ROOT}
+
+RUN apk update && \
+	apk add bash && \
+	apk add git
+
+COPY ./src ./src
+COPY go.mod .
+COPY go.sum .
+RUN go mod tidy
```

ここはDockerのキャッシュ効率を意識しましょう！

```suggestion
COPY go.mod go.sum .
RUN go mod download
COPY ./src ./src
```

とすると良いです！




DockerのキャッシュというのはCOPYファイルに変更があった場合、基本的にそれ以降の行のキャッシュを破棄して全部ビルドするようになっています。

つまり8行目のRUNを先に行うのと同様にgo.modやsumはあまり変更が加わりにくく、またsrcはコードなので変更が加わりやすいので、先にgo mod downloadをしてライブラリ群をダウンロードしてキャッシュしておくことで、びるどがはやくなります。

またgo mod tidyは特にしなくて良いです！

---

おっしゃる通り、modやsumの変更は基本的に少ないですね...。

## メモ
- [ADD or COPY](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy)
- [Docker Buildにおけるリードタイム短縮のための3つの改善ポイント](https://tech.plaid.co.jp/improve_docker_build_efficiency)